backend/src/schema/resolvers/registration.spec.js

```diff
@@ -0,0 +1,402 @@
+        })
+
+        it('creates no additional invitation codes', async done => {
+          try {
+            await action()
+          } catch (e) {
+            const invitationCodes = await instance.all('InvitationCode')
+            await expect(invitationCodes.toJson()).resolves.toHaveLength(3)
+            done()
+          }
+        })
+      })
+
+      describe('as a strong donator', () => {
+        beforeEach(() => {
+          // What is the setup?
+        })
+
+        it.todo('can invite more people')
+        // it('can invite more people', async () => {
+        // await action()
+        // const invitationQuery = `{ User { createdAt } }`
+        // const { User: users } = await client.request(invitationQuery )
+        // expect(users).toHaveLength(3 + 1 + 1)
+        // })
+      })
+    })
+  })
+})
+
+describe('SignupByInvitation', () => {
+  const mutation = `mutation($email: String!, $token: String!) {
+    SignupByInvitation(email: $email, token: $token) { email }
+  }`
+
+  beforeEach(() => {
+    client = new GraphQLClient(host)
+    action = async () => {
+      return client.request(mutation, variables)
+    }
+  })
+
+  describe('with valid email but invalid InvitationCode', () => {
+    beforeEach(() => {
+      variables.email = 'any-email@example.org'
+      variables.token = 'wut?'
+    })
+
+    it('throws UserInputError', async () => {
+      await expect(action()).rejects.toThrow('Invitation code already used or does not exist.')
+    })
+  })
+
+  describe('with valid InvitationCode', () => {
+    beforeEach(async () => {
+      const inviterParams = {
+        name: 'Inviter',
+        email: 'inviter@example.org',
+        password: '1234',
+      }
+      const factory = Factory()
+      await factory.create('User', inviterParams)
+      const headersOfInviter = await login(inviterParams)
+      const anotherClient = new GraphQLClient(host, { headers: headersOfInviter })
+      const invitationMutation = `mutation { CreateInvitationCode { token } }`
+      const {
+        CreateInvitationCode: { token },
+      } = await anotherClient.request(invitationMutation)
+      variables.token = token
+    })
+
+    describe('given an invalid email', () => {
+      beforeEach(() => {
+        variables.email = 'someuser'
+      })
+
+      it('throws `email is not a valid email`', async () => {
+        await expect(action()).rejects.toThrow('"email" must be a valid email')
+      })
+
+      it('creates no EmailAddress node', async done => {
+        try {
+          await action()
+        } catch (e) {
+          const emailAddresses = await instance.all('EmailAddress')
+          expect(emailAddresses).toHaveLength(0)
+          done()
+        }
+      })
+    })
+
+    describe('given a valid email', () => {
+      beforeEach(() => {
+        variables.email = 'someUser@example.org'
+      })
+
+      it('resolves', async () => {
+        await expect(action()).resolves.toEqual({
```

Comment:
> [<img alt="mattwr18" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/mattwr18) **Authored by [mattwr18](https://github.com/mattwr18)**
_<time datetime="2019-07-03T18:14:38Z" title="Wednesday, July 3rd 2019, 8:14:38 pm +02:00">Jul 3, 2019</time>_

---
> **Outdated (history rewrite)** - original diff
---
```diff
@@ -0,0 +1,395 @@
+import { GraphQLClient } from 'graphql-request'
+import Factory from '../../seed/factories'
+import { host, login } from '../../jest/helpers'
+import { neode } from '../../bootstrap/neo4j'
+
+let factory
+let client
+let variables
+let action
+let userParams
+const instance = neode()
+
+beforeEach(async () => {
+  variables = {}
+  factory = Factory()
+})
+
+afterEach(async () => {
+  await factory.cleanDatabase()
+})
+
+describe('CreateInvitationCode', () => {
+  const mutation = `mutation { CreateInvitationCode { token } }`
+
+  it('throws Authorization error', async () => {
+    const client = new GraphQLClient(host)
+    await expect(client.request(mutation)).rejects.toThrow('Not Authorised!')
+  })
+
+  describe('authenticated', () => {
+    beforeEach(async () => {
+      userParams = {
+        id: 'i123',
+        name: 'Inviter',
+        email: 'inviter@example.org',
+        password: '1234',
+      }
+      action = async () => {
+        const factory = Factory()
+        await factory.create('User', userParams)
+        const headers = await login(userParams)
+        client = new GraphQLClient(host, { headers })
+        return client.request(mutation)
+      }
+    })
+
+    it('resolves', async () => {
+      await expect(action()).resolves.toEqual({
+        CreateInvitationCode: { token: expect.any(String) },
+      })
+    })
+
+    it('creates an InvitationCode with a `createdAt` attribute', async () => {
+      await action()
+      const codes = await instance.all('InvitationCode')
+      const invitation = await codes.first().toJson()
+      expect(invitation.createdAt).toBeTruthy()
+      expect(Date.parse(invitation.createdAt)).toEqual(expect.any(Number))
+    })
+
+    it('relates inviting User to InvitationCode', async () => {
+      await action()
+      const result = await instance.cypher(
+        'MATCH(code:InvitationCode)<-[:GENERATED]-(user:User) RETURN user',
+      )
+      const inviter = instance.hydrateFirst(result, 'user', instance.model('User'))
+      await expect(inviter.toJson()).resolves.toEqual(expect.objectContaining({ name: 'Inviter' }))
+    })
+
+    describe('who has invited a lot of users already', () => {
+      beforeEach(() => {
+        action = async () => {
+          const factory = Factory()
+          await factory.create('User', userParams)
+          const headers = await login(userParams)
+          client = new GraphQLClient(host, { headers })
+          await Promise.all(
+            [1, 2, 3].map(() => {
+              return client.request(mutation)
+            }),
+          )
+          return client.request(mutation, variables)
+        }
+      })
+
+      describe('as ordinary `user`', () => {
+        it('throws `Not Authorised` because of maximum number of invitations', async () => {
+          await expect(action()).rejects.toThrow('Not Authorised')
+        })
+
+        it('creates no additional invitation codes', async done => {
+          try {
+            await action()
+          } catch (e) {
+            const invitationCodes = await instance.all('InvitationCode')
+            await expect(invitationCodes.toJson()).resolves.toHaveLength(3)
+            done()
+          }
+        })
+      })
+
+      describe('as a strong donator', () => {
+        beforeEach(() => {
+          // What is the setup?
+        })
+
+        it.todo('can invite more people')
+        // it('can invite more people', async () => {
+        // await action()
+        // const invitationQuery = `{ User { createdAt } }`
+        // const { User: users } = await client.request(invitationQuery )
+        // expect(users).toHaveLength(3 + 1 + 1)
+        // })
+      })
+    })
+  })
+})
+
+describe('SignupByInvitation', () => {
+  const mutation = `mutation($email: String!, $token: String!) {
+    SignupByInvitation(email: $email, token: $token) { email }
+  }`
+
+  beforeEach(() => {
+    client = new GraphQLClient(host)
+    action = async () => {
+      return client.request(mutation, variables)
+    }
+  })
+
+  it.todo('throws Authorization error')
+
+  describe('with invalid InvitationCode', () => {
+    beforeEach(() => {
+      variables.token = 'wut?'
+    })
+
+    it.todo('throws UserInputError')
+  })
+
+  describe('with valid InvitationCode', () => {
+    beforeEach(async () => {
+      const inviterParams = {
+        name: 'Inviter',
+        email: 'inviter@example.org',
+        password: '1234',
+      }
+      const factory = Factory()
+      await factory.create('User', inviterParams)
+      const headersOfInviter = await login(inviterParams)
+      const anotherClient = new GraphQLClient(host, { headers: headersOfInviter })
+      const invitationMutation = `mutation { CreateInvitationCode { token } }`
+      const {
+        CreateInvitationCode: { token },
+      } = await anotherClient.request(invitationMutation)
+      variables.token = token
+    })
+
+    describe('given an invalid email', () => {
+      beforeEach(() => {
+        variables.email = 'someuser'
+      })
+
+      it('throws `email is not a valid email`', async () => {
+        await expect(action()).rejects.toThrow('"email" must be a valid email')
+      })
+
+      it('creates no EmailAddress node', async done => {
+        try {
+          await action()
+        } catch (e) {
+          const emailAddresses = await instance.all('EmailAddress')
+          expect(emailAddresses).toHaveLength(0)
+          done()
+        }
+      })
+    })
+
+    describe('given a valid email', () => {
+      beforeEach(() => {
+        variables.email = 'someUser@example.org'
+      })
+
+      it('resolves', async () => {
+        await expect(action()).resolves.toEqual({
```


I don't want to be pedantic, and I acknowledge that we each have different styles, but it takes me a little longer to verify what is happening here because the variables are all named the same `action` in this case does what? I need to scroll up to check.
consider naming it `signupByInvitationAction`, etc...